src/main/scala/io/picnicml/doddlemodel/preprocessing/Binarizer.scala

```diff
@@ -0,0 +1,38 @@
+package io.picnicml.doddlemodel.preprocessing
+
+import breeze.linalg.{*, DenseVector}
+import io.picnicml.doddlemodel.data.Feature.FeatureIndex
+import io.picnicml.doddlemodel.data.{Features, RealVector}
+import io.picnicml.doddlemodel.typeclasses.Transformer
+
+case class Binarizer private (private val featureIndex: FeatureIndex,
```

Comment:
It's a good point, I actually put it like that originally, but there was a compilation error because the case class auto-generates an `apply` method with same signature as the one I create inside `Binarizer` companion object. Here's the copy-paste of the error from CI:
```
[info] Non-compiled module 'compiler-bridge_2.11' for Scala 2.11.12. Compiling...
[info]   Compilation completed in 7.738s.
[error] /home/circleci/repo/src/main/scala/io/picnicml/doddlemodel/preprocessing/Binarizer.scala:8:12: method apply is defined twice
[error]   conflicting symbols both originated in file '/home/circleci/repo/src/main/scala/io/picnicml/doddlemodel/preprocessing/Binarizer.scala'
[error] case class Binarizer private (private val thresholds: RealVector,
[error]            ^
[error] one error found
[error] (Compile / compileIncremental) Compilation failed
```

Do you have an idea on how to fix this problem in a nicer way?

Reply:
Hmm weird, this compiles for me:
```scala
case class Binarizer private (private val thresholds: RealVector, private val featureIndex: FeatureIndex)

object Binarizer {

  def apply(threshold: Double, featureIndex: FeatureIndex): Binarizer = {
    val numNumeric: Int = featureIndex.numerical.columnIndices.length
    require(numNumeric > 0, "There must be at least 1 numeric column in the given data")
    val thresholdsExtended = DenseVector.fill(numNumeric) {threshold}
    Binarizer(thresholdsExtended, featureIndex)
  }

  def apply(thresholds: RealVector, featureIndex: FeatureIndex): Binarizer = {
    val numNumeric = featureIndex.numerical.columnIndices.length
    require(numNumeric > 0, "There must be at least 1 numeric column in the given data")
    require(numNumeric == thresholds.length, "A threshold should be given for every numerical column")
    Binarizer(thresholds, featureIndex)
  }
  ...
}
```
Btw no need to put definitions in multiple lines if they fit in 120 characters, take a look at [this](https://github.com/picnicml/doddle-model#development).

Reply:
You are right about that, it does compile. However, making the above change introduces a problem - some Binarizer tests do not pass (or end, for that matter). The problem seems to be that the `Binarizer(thresholds, featureIndex)` call in the second `apply` function calls itself instead of the `apply` method that is automatically created by the case class (as they have same signatures).

Introducing the `new` keyword (redundant, but I thought it could fix things) fixes the problem, but does not compile on Scala 2.11.12:
```
[info] Compiling 48 Scala sources to /home/matej/Documents/doddle-model/target/scala-2.11/classes ...
[error] /home/matej/Documents/doddle-model/src/main/scala/io/picnicml/doddlemodel/preprocessing/Binarizer.scala:16:5: ambiguous reference to overloaded definition,
[error] both method apply in object Binarizer of type (thresholds: io.picnicml.doddlemodel.data.RealVector, featureIndex: io.picnicml.doddlemodel.data.Feature.FeatureIndex)io.picnicml.doddlemodel.preprocessing.Binarizer
[error] and  method apply in object Binarizer of type (thresholds: io.picnicml.doddlemodel.data.RealVector, featureIndex: io.picnicml.doddlemodel.data.Feature.FeatureIndex)io.picnicml.doddlemodel.preprocessing.Binarizer
[error] match argument types (breeze.linalg.DenseVector[Double],io.picnicml.doddlemodel.data.Feature.FeatureIndex) and expected result type io.picnicml.doddlemodel.preprocessing.Binarizer
[error]     Binarizer(thresholdsExtended, featureIndex)
[error]     ^
[error] /home/matej/Documents/doddle-model/src/main/scala/io/picnicml/doddlemodel/preprocessing/Binarizer.scala:8:12: method apply is defined twice
[error]   conflicting symbols both originated in file '/home/matej/Documents/doddle-model/src/main/scala/io/picnicml/doddlemodel/preprocessing/Binarizer.scala'
[error] case class Binarizer private (private val thresholds: RealVector, private val featureIndex: FeatureIndex)
[error]            ^
[error] two errors found
```

Reply:
Oh I see what the problem is. In that case, it makes sense to simply make the constructor public and remove the "overloaded" apply method:
```scala
case class Binarizer(private val thresholds: RealVector, private val featureIndex: FeatureIndex)

object Binarizer {
  // only keep this apply method
  def apply(threshold: Double, featureIndex: FeatureIndex): Binarizer = {
    val numNumeric: Int = featureIndex.numerical.columnIndices.length
    require(numNumeric > 0, "There must be at least 1 numeric column in the given data")
    val thresholdsExtended = DenseVector.fill(numNumeric) {threshold}
    Binarizer(thresholdsExtended, featureIndex)
  }
 ...
}
```
If after refactoring you still need to validate the input arguments you can do that directly in the case class definition:
```scala
case class Binarizer(private val thresholds: RealVector, private val featureIndex: FeatureIndex) {
  require(something > 0, "'something' has to be positive.")
}
```